### ISSUE TITLE
Aborted upload causes server crash

### ISSUE THREAD

**Issue by nitzo** [balderdashy/skipper#74]
I am attempting to build a new adapter and while determining the correct way to respond to errors during upload I encountered an issue that when an upload is aborted for some reason  the server crashes. 

NOTE: This also happens with the built in skipper-disk adapter.

This can happen both by user (browser) abort or by adapter abort  (by emitting 'error' or by calling the done() callback in _writeFile).

After some debugging I discovered that skipper provides an error handler for these cases in prototype.writeFile.js. This handler is registered with the 'once' option meaning it will only be called once.

I think what happens is that the error is emitted twice.

Once from the current upstream that is aborted (Because of an error, for example closing the browser tab will close the socket and an error to be emitted) and once from multiparty when the request is aborted (See index.js in multiparty line 196.) this happens later after you send a response in the callback provided to skipper.upload method.

I tried not to send a response to the client and the server did not crash.

Changing the error handler in skipper from 'once' to 'on' also solved the problem although I am not sure the error processing there is correct if it is called twice. I did not dig in further into skipper so I do not fully understand it's internals.

Can someone shed some light on this?

I'll be more than happy to provide a fix given I know what's the correct way to handle the error.

Thanks!

**Comment by ecfaria** [comment]
The same is happening to me.

**Comment by devinivy** [comment]
I have a hunch that the stream needs to be ended there.
http://www.bennadel.com/blog/2692-you-have-to-explicitly-end-streams-after-pipes-break-in-node-js.htm

**Comment by devinivy** [comment]
You can also run that stream in a domain, and let the domain catch your unhandled error: http://stackoverflow.com/questions/21771220/error-handling-with-node-js-streams

**Comment by tjwebb** [comment]
@devinivy any thoughts on how we can drop on a quick fix?

**Comment by devinivy** [comment]
A quick fix?  I expect this would do! (ref [here](https://github.com/balderdashy/skipper/blob/master/standalone/Upstream/prototype.writeFile.js#L31))

```javascript
module.exports = function writeFile (__filestream) {
    /* ... */
    __filestream.on('error', (function(err) {

            var hasFired = false;

            return function(err) {
                if (hasFired) return;
                hasFired = true;
                /* All the original callback code */
            }
    })()
    /* ... */
}
```

Basically, use `on` instead of `once` (because you don't want to unregister this event handler!) then keep track manually whether the callback has fired or not.

**Comment by cburatto** [comment]
Should reopen this?
https://github.com/balderdashy/skipper/issues/49

**Comment by nitzo** [comment]
I have submitted a merge request for the proposed solution. 
I don't think ending the stream will help here since to the best of my understanding multiparty fires an error event no matter what.

Thanks @devinivy !

**Comment by devinivy** [comment]
Sure thing!  I hope it works-out for now.  It seems like there's more to understand and clean-up here.

**Comment by devyn** [comment]
I'm getting this same error with skipper-s3 if the browser request is aborted while sending the file to S3.

**Comment by metzuda** [comment]
I'm still having issue after trying @nitzo fix from pull request #78, but only if I am not using Chrome.

**Comment by sailsbot** [comment]
Thanks for posting, @nitzo.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by megakoresh** [comment]
Having this with the azure adapter. This is a very serious issue, is anyone still looking into this?

**Comment by Uminily** [comment]
Same issue with S3 Adapter.

**Comment by Emiyashirou** [comment]
Still have this issue with skipper-s3 adapter.

**Comment by boshido** [comment]
Please fix this soon
It still have this issue with skipper-s3 adapter.

**Comment by metzuda** [comment]
+!

**Comment by mikermcneil** [comment]
@boshido @metzuda @Emiyashirou @Uminily @megakoresh @devinivy I reopened https://github.com/balderdashy/skipper/issues/49 -- would you guys please let us know there if this is still happening for you?  Also, if you are still seeing this, when you respond to the other issue, if you could provide [version information](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#ii-issue-contributions) for both Skipper and the skipper adapter you're using, as well as whether you're using Sails vs. Express (and which version), as well as your node version, that would help a lot here.  Thanks!
